src/systemcmds/gpio/gpio.cpp

```diff
@@ -0,0 +1,190 @@
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file gpio_test.cpp
+ * @author Julian Kent <julian@auterion.com>
+ *
+ * GPIO read and write tool
+ */
+
+#include <px4_platform_common/log.h>
+#include <px4_platform_common/module.h>
+#include <px4_platform_common/px4_config.h>
+
+
+__BEGIN_DECLS
+__EXPORT int gpio_main(int argc, char *argv[]);
+__END_DECLS
+
+
+static void
+usage(const char *reason)
+{
+	printf("FAIL\n");
+
+	if (reason != nullptr) {
+		PX4_WARN("%s", reason);
+	}
+
+	PRINT_MODULE_DESCRIPTION(
+		R"DESCR_STR(
+### Description
+This command is used to read and write GPIOs
+
+gpio read <PORT> <PIN> [PULLDOWN|PULLUP] [--force]
+gpio write <PORT> <PIN> <VALUE> [PUSHPULL|OPENDRAIN] [--force]
+
+### Examples
+Read the value on port H pin 4 configured as pullup, and it is high
+$ gpio_test read H 4 PU
+1 OK
+
+Set the output value on Port E pin 7 to high
+$ gpio_test write E 7 1 --force
+OK
+
+)DESCR_STR");
+
+}
+
+int
+gpio_main(int argc, char *argv[])
+{
+	if (argc < 4) {
+		usage("not enough arguments");
+		return -1;
+	}
+
+	const char* command = argv[1];
+	bool is_read = false;
+	uint32_t mask = 0;
+	if (strcmp(command, "read") == 0) {
+		is_read = true;
+		mask |= GPIO_INPUT;
+	} else if (strcmp(command, "write") == 0) {
+		if (argc < 5) {
+			usage("not enough arguments");
+			return -1;
+		}
+		mask |= GPIO_OUTPUT;
+	} else {
+		usage("command not read or write");
+		return -1;
+	}
+
+	const char* port_string = argv[2];
+	if (strlen(port_string) != 1) {
+		usage("port not a single character");
+	}
+	char port = port_string[0];
+	if ('A' <= port && port <= 'K') {
+		mask |= ((port - 'A') << GPIO_PORT_SHIFT) & GPIO_PORT_MASK;
+	} else if ('a' <= port && port <= 'k') {
+		mask |= ((port - 'a') << GPIO_PORT_SHIFT) & GPIO_PORT_MASK;
+	}else {
```

Comment:
Run this through astyle (`make format`)?

Reply:
I did, I was expecting it to fix things but it didn't even change spaces to tabs so I had to do that manually. I guess there's some issues in the `make format` where it doesn't touch certain folders?

Reply:
Did some debugging, `Tools/astyle/files_to_check_code_style.sh | grep gpio.cpp` gives me the file, so for some reason astyle isn't correcting it. I found another too, I'll fix manually.